.changeset/itchy-swans-travel.md

```diff
@@ -0,0 +1,5 @@
+---
+"@remix-run/router": patch
+---
+
+Fix explicit `replace` on submissions and `PUSH` on submission to new paths
```

Comment:
Change implementation to be based on the _next_ location (either from `formAction` or `redirect`).  Replace if it's the same (pathname + search), push otherwise.